File: src/main/java/me/lokka30/levelledmobs/LevelledMobs.java
/*
 * Copyright (c) 2020-2022  lokka30
 * Use of this source code is governed by the GNU AGPL v3.0
 * license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;
import java.util.Stack;
import java.util.TreeMap;
import java.util.WeakHashMap;
import me.lokka30.levelledmobs.commands.LevelledMobsCommand;
import me.lokka30.levelledmobs.customdrops.CustomDropsHandler;
import me.lokka30.levelledmobs.listeners.BlockPlaceListener;
import me.lokka30.levelledmobs.listeners.ChunkLoadListener;
import me.lokka30.levelledmobs.listeners.EntityDamageDebugListener;
import me.lokka30.levelledmobs.listeners.PlayerInteractEventListener;
import me.lokka30.levelledmobs.managers.LevelManager;
import me.lokka30.levelledmobs.managers.MobDataManager;
import me.lokka30.levelledmobs.managers.MobHeadManager;
import me.lokka30.levelledmobs.managers.MobsQueueManager;
import me.lokka30.levelledmobs.managers.NametagQueueManager;
import me.lokka30.levelledmobs.managers.PlaceholderApiIntegration;
import me.lokka30.levelledmobs.misc.FileLoader;
import me.lokka30.levelledmobs.wrappers.LivingEntityWrapper;
import me.lokka30.levelledmobs.misc.NamespacedKeys;
import me.lokka30.levelledmobs.misc.NametagTimerChecker;
import me.lokka30.levelledmobs.misc.YmlParsingHelper;
import me.lokka30.levelledmobs.nametag.Definitions;
import me.lokka30.levelledmobs.nametag.ServerVersionInfo;
import me.lokka30.levelledmobs.rules.RulesManager;
import me.lokka30.levelledmobs.rules.RulesParsingManager;
import me.lokka30.levelledmobs.util.ConfigUtils;
import me.lokka30.levelledmobs.util.QuickTimer;
import me.lokka30.levelledmobs.util.Utils;
import org.bukkit.Bukkit;
import org.bukkit.command.CommandSender;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.LivingEntity;
import org.bukkit.event.HandlerList;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;

/**
 * This is the main class of the plugin. Bukkit will call onLoad and onEnable on startup, and
 * onDisable on shutdown.
 *
 * @author lokka30, stumper66
 * @since 1.0
 */
public final class LevelledMobs extends JavaPlugin {

    public LevelInterface levelInterface;
    public LevelManager levelManager;
    public MobDataManager mobDataManager;
    public CustomDropsHandler customDropsHandler;
    public ChunkLoadListener chunkLoadListener;
    public BlockPlaceListener blockPlaceListener;
    public PlayerInteractEventListener playerInteractEventListener;
    public NamespacedKeys namespacedKeys;
    public Companion companion;
    public MobHeadManager mobHeadManager;
    public RulesParsingManager rulesParsingManager;
    public RulesManager rulesManager;
    public MobsQueueManager mobsQueueManager;
    public NametagQueueManager nametagQueueManager;
    public NametagTimerChecker nametagTimerChecker;
    public final Object attributeSyncObject = new Object();
    public LevelledMobsCommand levelledMobsCommand;
    public Random random;
    public PlaceholderApiIntegration placeholderApiIntegration;
    public boolean migratedFromPre30;
    public YmlParsingHelper helperSettings;
    public long playerLevellingMinRelevelTime;
    public int maxPlayersRecorded;
    private Definitions definitions;
    private ServerVersionInfo ver;
    private static LevelledMobs instance;

    // Configuration
    public YamlConfiguration settingsCfg;
    public YamlConfiguration messagesCfg;
    public YamlConfiguration dropsCfg;
    public final ConfigUtils configUtils = new ConfigUtils(this);

    // Misc
    public Map<String, Set<String>> customMobGroups;
    public EntityDamageDebugListener entityDamageDebugListener;
    private long loadTime;
    public WeakHashMap<LivingEntity, Instant> playerLevellingEntities;
    public Stack<LivingEntityWrapper> cacheCheck;

    @Override
    public void onLoad() {
        instance = this;
    }

    @Override
    public void onEnable() {
        final QuickTimer timer = new QuickTimer();

        this.ver = new ServerVersionInfo();
        this.definitions = new Definitions();
        this.nametagQueueManager = new NametagQueueManager(this);
        this.mobsQueueManager = new MobsQueueManager(this);
        this.companion = new Companion(this);
        this.mobDataManager = new MobDataManager(this);
        this.mobHeadManager = new MobHeadManager(this);
        this.rulesParsingManager = new RulesParsingManager(this);
        this.rulesManager = new RulesManager(this);
        this.nametagTimerChecker = new NametagTimerChecker(this);
        this.namespacedKeys = new NamespacedKeys(this);
        this.playerLevellingEntities = new WeakHashMap<>();
        this.helperSettings = new YmlParsingHelper();
        this.random = new Random();
        this.customMobGroups = new TreeMap<>();
        this.levelInterface = new LevelManager(this);
        if (!companion.loadFiles(false)) {
            // had fatal error reading required files
            Bukkit.getPluginManager().disablePlugin(this);
            return;
        }
        definitions.useTranslationComponents = helperSettings.getBoolean(settingsCfg,
                "use-translation-components", true);
        definitions.setUseLegacySerializer(helperSettings.getBoolean(settingsCfg,
                "use-legacy-serializer", true));
        nametagQueueManager.nametagSenderHandler.refresh();
        companion.registerListeners();
        companion.registerCommands();

        Utils.logger.info("Running misc procedures");
        if (this.nametagQueueManager.hasNametagSupport()) {
            levelManager.startNametagAutoUpdateTask();
            levelManager.startNametagTimer();
        }

        prepareToLoadCustomDrops();
        companion.startCleanupTask();
        companion.setupMetrics();
        companion.checkUpdates();

        loadTime += timer.getTimer();
        Utils.logger.info("Start-up complete (took " + loadTime + "ms)");
    }

    private void prepareToLoadCustomDrops(){
        if (Bukkit.getPluginManager().getPlugin("LM_Items") != null){
            final LevelledMobs mainInstance = this;
            Bukkit.getScheduler().scheduleSyncDelayedTask(this, () -> customDropsHandler.customDropsParser.loadDrops(
                    FileLoader.loadFile(mainInstance, "customdrops", FileLoader.CUSTOMDROPS_FILE_VERSION)
            ), 10L);
        }
        else{
            customDropsHandler.customDropsParser.loadDrops(
                    FileLoader.loadFile(this, "customdrops", FileLoader.CUSTOMDROPS_FILE_VERSION)
            );
        }
    }

    public void reloadLM(final @NotNull CommandSender sender) {
        migratedFromPre30 = false;
        List<String> reloadStartedMsg = messagesCfg.getStringList(
            "command.levelledmobs.reload.started");
        reloadStartedMsg = Utils.replaceAllInList(reloadStartedMsg, "%prefix%",
            configUtils.getPrefix());
        reloadStartedMsg = Utils.colorizeAllInList(reloadStartedMsg);
        reloadStartedMsg.forEach(sender::sendMessage);

        companion.loadFiles(true);

        List<String> reloadFinishedMsg = messagesCfg.getStringList(
            "command.levelledmobs.reload.finished");
        reloadFinishedMsg = Utils.replaceAllInList(reloadFinishedMsg, "%prefix%",
            configUtils.getPrefix());
        reloadFinishedMsg = Utils.colorizeAllInList(reloadFinishedMsg);

        if (helperSettings.getBoolean(settingsCfg, "debug-entity-damage")
            && !configUtils.debugEntityDamageWasEnabled) {
            configUtils.debugEntityDamageWasEnabled = true;
            Bukkit.getPluginManager().registerEvents(entityDamageDebugListener, this);
        } else if (!helperSettings.getBoolean(settingsCfg, "debug-entity-damage")
            && configUtils.debugEntityDamageWasEnabled) {
            configUtils.debugEntityDamageWasEnabled = false;
            HandlerList.unregisterAll(entityDamageDebugListener);
        }

        if (helperSettings.getBoolean(settingsCfg, "ensure-mobs-are-levelled-on-chunk-load")
            && !configUtils.chunkLoadListenerWasEnabled) {
            configUtils.chunkLoadListenerWasEnabled = true;
            Bukkit.getPluginManager().registerEvents(chunkLoadListener, this);
        } else if (!helperSettings.getBoolean(settingsCfg, "ensure-mobs-are-levelled-on-chunk-load")
            && configUtils.chunkLoadListenerWasEnabled) {
            configUtils.chunkLoadListenerWasEnabled = false;
            HandlerList.unregisterAll(chunkLoadListener);
        }

        levelManager.entitySpawnListener.processMobSpawns = helperSettings.getBoolean(settingsCfg,
            "level-mobs-upon-spawn", true);

        levelManager.clearRandomLevellingCache();
        configUtils.playerLevellingEnabled = rulesManager.isPlayerLevellingEnabled();
        rulesManager.clearTempDisabledRulesCounts();
        definitions.useTranslationComponents = helperSettings.getBoolean(settingsCfg,
                "use-translation-components", true);
        definitions.setUseLegacySerializer(helperSettings.getBoolean(settingsCfg,
                "use-legacy-serializer", true));
        nametagQueueManager.nametagSenderHandler.refresh();

        reloadFinishedMsg.forEach(sender::sendMessage);
    }

    public static LevelledMobs getInstance(){
        return instance;
    }

    public @NotNull Definitions getDefinitions(){
        return this.definitions;
    }

    public @NotNull ServerVersionInfo getVerInfo() { return this.ver; }

    @Override
    public void onDisable() {
        final QuickTimer disableTimer = new QuickTimer();
        disableTimer.start();

        levelManager.stopNametagAutoUpdateTask();
        companion.shutDownAsyncTasks();

        Utils.logger.info("Shut-down complete (took " + disableTimer.getTimer() + "ms)");
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/LevelledMobs.kt
package io.github.arcaneplugins.levelledmobs

import io.github.arcaneplugins.levelledmobs.commands.CommandHandler
import io.github.arcaneplugins.levelledmobs.customdrops.CustomDropsHandler
import io.github.arcaneplugins.levelledmobs.debug.DebugManager
import io.github.arcaneplugins.levelledmobs.listeners.BlockPlaceListener
import io.github.arcaneplugins.levelledmobs.listeners.ChunkLoadListener
import io.github.arcaneplugins.levelledmobs.listeners.EntityDamageDebugListener
import io.github.arcaneplugins.levelledmobs.listeners.EntityDamageListener
import io.github.arcaneplugins.levelledmobs.listeners.EntityDeathListener
import io.github.arcaneplugins.levelledmobs.listeners.EntityTransformListener
import io.github.arcaneplugins.levelledmobs.listeners.PlayerDeathListener
import io.github.arcaneplugins.levelledmobs.listeners.PlayerInteractEventListener
import io.github.arcaneplugins.levelledmobs.managers.LevelManager
import io.github.arcaneplugins.levelledmobs.managers.MobDataManager
import io.github.arcaneplugins.levelledmobs.managers.MobsQueueManager
import io.github.arcaneplugins.levelledmobs.managers.NametagQueueManager
import io.github.arcaneplugins.levelledmobs.managers.NotifyManager
import io.github.arcaneplugins.levelledmobs.managers.PlaceholderApiIntegration
import io.github.arcaneplugins.levelledmobs.misc.NametagTimerChecker
import io.github.arcaneplugins.levelledmobs.misc.YmlParsingHelper
import io.github.arcaneplugins.levelledmobs.nametag.Definitions
import io.github.arcaneplugins.levelledmobs.nametag.ServerVersionInfo
import io.github.arcaneplugins.levelledmobs.rules.RulesManager
import io.github.arcaneplugins.levelledmobs.rules.RulesParser
import io.github.arcaneplugins.levelledmobs.rules.strategies.RandomLevellingStrategy
import io.github.arcaneplugins.levelledmobs.util.ConfigUtils
import io.github.arcaneplugins.levelledmobs.util.Log
import io.github.arcaneplugins.levelledmobs.util.MessageUtils
import io.github.arcaneplugins.levelledmobs.util.QuickTimer
import io.github.arcaneplugins.levelledmobs.util.Utils
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper
import io.github.arcaneplugins.levelledmobs.wrappers.SchedulerWrapper
import java.time.Instant
import java.util.Random
import java.util.Stack
import java.util.WeakHashMap
import java.util.function.Consumer
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.event.HandlerList
import org.bukkit.plugin.java.JavaPlugin


/**
 * This is the main class of the plugin. Bukkit will call onLoad and onEnable on startup, and
 * onDisable on shutdown.
 *
 * @author lokka30, stumper66
 * @since 1.0
 */
class LevelledMobs : JavaPlugin() {
    val levelInterface: LevelInterface2 = LevelManager()
    val levelManager = LevelManager()
    val mobDataManager = MobDataManager()
    val customDropsHandler = CustomDropsHandler()
    val chunkLoadListener = ChunkLoadListener()
    val blockPlaceListener = BlockPlaceListener()
    val playerInteractEventListener = PlayerInteractEventListener()
    val playerDeathListener = PlayerDeathListener()
    val entityDamageListener = EntityDamageListener()
    val entityTransformListener = EntityTransformListener()
    val entityDeathListener = EntityDeathListener()
    val mainCompanion = MainCompanion()
    val rulesParsingManager = RulesParser()
    val rulesManager = RulesManager()
    val mobsQueueManager = MobsQueueManager()
    val nametagQueueManager = NametagQueueManager()
    val nametagTimerChecker = NametagTimerChecker()
    val attributeSyncObject = Any()
    val random = Random()
    var placeholderApiIntegration: PlaceholderApiIntegration? = null
        internal set
    val helperSettings = YmlParsingHelper(YamlConfiguration())
    var playerLevellingMinRelevelTime = 0L
        internal set
    var maxPlayersRecorded = 0
    val debugManager = DebugManager()
    val definitions = Definitions()
    val ver = ServerVersionInfo()

    // Configuration
    var messagesCfg = YamlConfiguration()
        internal set
    val configUtils = ConfigUtils()

    // Misc
    val customMobGroups = mutableMapOf<String, MutableSet<String>>()
    var entityDamageDebugListener = EntityDamageDebugListener()
    private var loadTime = 0L
    val playerLevellingEntities = WeakHashMap<LivingEntity, Instant>()
    var cacheCheck: Stack<LivingEntityWrapper>? = null

    companion object {
        @JvmStatic
        lateinit var instance: LevelledMobs
            private set
    }

    override fun onLoad() {
        instance = this
        CommandHandler.load(CommandHandler.LoadingStage.ON_LOAD)
    }

    override fun onEnable() {
        val timer = QuickTimer()

        CommandHandler.load(CommandHandler.LoadingStage.ON_ENABLE)
        this.ver.load()
        if (ver.minecraftVersion <= 1.18){
            Log.sev("This minecraft version is NOT supported. Use at your own risk!")
        }
        this.definitions.load()
        this.nametagQueueManager.load()
        this.mainCompanion.load()
        (this.levelInterface as LevelManager).load()
        if (!mainCompanion.loadFiles()) {
            // had fatal error reading required files
            Bukkit.getPluginManager().disablePlugin(this)
            return
        }
        definitions.useTranslationComponents = helperSettings.getBoolean(
            "use-translation-components", true
        )
        definitions.setUseLegacySerializer(
            helperSettings.getBoolean("use-legacy-serializer", true)
        )
        nametagQueueManager.nametagSenderHandler.refresh()
        mainCompanion.registerListeners()

        Log.inf("Running misc procedures")
        if (nametagQueueManager.hasNametagSupport) {
            levelManager.startNametagAutoUpdateTask()
            levelManager.startNametagTimer()

            if (!ver.isRunningFolia){
                val scheduler = SchedulerWrapper {
                    nametagQueueManager.taskChecker()
                    mobsQueueManager.taskChecker()
                }
                scheduler.runTaskTimerAsynchronously(50000, 5000)
            }
        }

        prepareToLoadCustomDrops()
        mainCompanion.checkSettingsWithMaxPlayerOptions()
        mainCompanion.startCleanupTask()
        mainCompanion.setupMetrics()
        mainCompanion.checkUpdates()

        loadTime += timer.timer
        Log.inf("Start-up complete (took ${loadTime}ms)")
    }

    override fun onDisable() {
        val disableTimer = QuickTimer()
        disableTimer.start()

        CommandHandler.load(CommandHandler.LoadingStage.ON_DISABLE)
        levelManager.stopNametagAutoUpdateTask()
        mainCompanion.shutDownAsyncTasks()

        Log.inf("Shut-down complete (took ${disableTimer.timer}ms)")
    }

    private fun prepareToLoadCustomDrops(){
        if (Bukkit.getPluginManager().getPlugin("LM_Items") == null &&
            mainCompanion.showCustomDrops) {
            customDropsHandler.customDropsParser.showCustomDropsDebugInfo(null)
        }
    }

    fun reloadLM(sender: CommandSender) {
        NotifyManager.clearLastError()
        customDropsHandler.customDropsParser.invalidExternalItems.clear()
        var reloadStartedMsg = messagesCfg.getStringList(
            "command.levelledmobs.reload.started"
        )
        reloadStartedMsg = Utils.replaceAllInList(
            reloadStartedMsg, "%prefix%",
            configUtils.prefix
        )
        reloadStartedMsg = Utils.colorizeAllInList(reloadStartedMsg)
        reloadStartedMsg.forEach(Consumer { s: String ->
            sender.sendMessage(s)
        })

        mainCompanion.reloadSender = sender
        mainCompanion.loadFiles()
        mainCompanion.checkListenersWithVariablePriorities()
        chunkLoadListener.load()
        RandomLevellingStrategy.clearCache()

        var reloadFinishedMsg = messagesCfg.getStringList(
            "command.levelledmobs.reload.finished"
        )
        reloadFinishedMsg = Utils.replaceAllInList(
            reloadFinishedMsg, "%prefix%",
            configUtils.prefix
        )
        reloadFinishedMsg = Utils.colorizeAllInList(reloadFinishedMsg)

        if (helperSettings.getBoolean("ensure-mobs-are-levelled-on-chunk-load")
            && !configUtils.chunkLoadListenerWasEnabled
        ) {
            configUtils.chunkLoadListenerWasEnabled = true
            Bukkit.getPluginManager().registerEvents(chunkLoadListener, this)
        } else if (!helperSettings.getBoolean( "ensure-mobs-are-levelled-on-chunk-load")
            && configUtils.chunkLoadListenerWasEnabled
        ) {
            configUtils.chunkLoadListenerWasEnabled = false
            HandlerList.unregisterAll(chunkLoadListener)
        }

        if (customDropsHandler.customDropsParser.hadParsingError && sender is Player){
            sender.sendMessage(MessageUtils.colorizeAll(
                "&b&lLevelledMobs:&r &6There was an error parsing customdrops.yml&r\n" +
                "Check the console log for more details"
            ))
        }

        rulesManager.clearTempDisabledRulesCounts()
        definitions.useTranslationComponents = helperSettings.getBoolean(
            "use-translation-components", true
        )
        definitions.setUseLegacySerializer(
            helperSettings.getBoolean(
                "use-legacy-serializer", true
            )
        )
        mainCompanion.checkSettingsWithMaxPlayerOptions()
        nametagQueueManager.nametagSenderHandler.refresh()

        reloadFinishedMsg.forEach(Consumer { s: String ->
            sender.sendMessage(s)
        })
    }
}